### PR TITLE
Improve ability to use negative numbers in filter & post-filter panel for query builder

### LINKED ISSUE #1980: Query builder filters with negative numbers reset to zero
Currently, when users use the filter panel in the query builder and input negative numbers, although running the query retrieves the correct results at first, the negative number defaults back to '0' when users click 'Edit Query in PURE' or when they save the query and attempt to load it back again. 

This same problem exists for the post-filter panel as well.

https://user-images.githubusercontent.com/73408381/222758891-f3025919-52de-490a-9404-620f4681ce54.mov

Kapture 2023-03-06 at 15 27 29

_Originally posted by @YannanGao-gs in https://github.com/finos/legend-studio/issues/1968#issuecomment-1453693644_

### PR DESCRIPTION
**Pull request by xannem**
## Summary

fixes #1980

Previously, the only way to include a negative number in the filter panel was for users to manually move the cursor to the beginning of the '0' number, press '-', and then move the cursor after the 0, type the number they wanted to be negative, and then delete the 0. This is very unintuitive for the user and a strange method they have to follow in order to write negative numbers. 

To resolve this, we change the input type to be `text` to allow the usage of `.` and `-` to show for the input while the user is typing, but when they click away we change it to be a number to keep our expected behavior and also improve UX for users trying to filter with negative numbers. 

Kapture 2023-03-01 at 15 49 22

Below is a video of the way originally to get negative numbers (pressing the negative sign if you have '0' will result in no activity) 
Kapture 2023-03-01 at 15 57 54

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

**Comment by YannanGao-gs** [comment]
I think there might be a bug. See video below

https://user-images.githubusercontent.com/73408381/222758891-f3025919-52de-490a-9404-620f4681ce54.mov

**Comment by MauricioUyaguari** [comment]
Tests ? :)

**Comment by xannem** [comment]
> I think there might be a bug. See video below

whoa you're right I checked it this is preexisting thing already from how we handle negative nums in filter good catch I will go add new issue for that

File: .changeset/common-starling-yodels.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@finos/legend-query-builder': patch
+---
+
+Add tests for expected behavior for entry of negative numbers
```
Comment (akphi) [inline .changeset/common-starling-yodels.md:5]: don't add changelog when you just add a test please.

File: packages/legend-graph/src/graph/MetaModelConst.ts [hunk 2]
```diff
@@ -30,6 +30,7 @@ export const PARSER_SECTION_MARKER = '###';
 export const FUNCTION_SIGNATURE_MULTIPLICITY_INFINITE_TOKEN = 'MANY';
 export const DIRECTORY_PATH_DELIMITER = '/';
 export const SOURCE_INFORMATION_PROPERTY_KEY_SUFFIX = 'sourceInformation';
+export const MINUS_STRING = 'minus';
```
Comment (akphi) [inline packages/legend-graph/src/graph/MetaModelConst.ts:33]: this is handled in query-builder, there's a list of supported functions over there, please move it there accordingly, and it should be called `QUERY_BUILDER_SUPPORTED_FUNCTIONS.MINUS` instead

File: packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx [hunk 3]
```diff
@@ -361,13 +361,82 @@ const NumberPrimitiveInstanceValueEditor = observer(
...
+      }
+
       let inputVal = isInteger
         ? parseInt(event.target.value, 10)
         : parseFloat(event.target.value);
       inputVal = isNaN(inputVal) ? 0 : inputVal;
-      instanceValue_setValue(valueSpecification, inputVal, 0);
+
+      valueIsNegativeSymbol
+        ? instanceValue_setValue(valueSpecification, NEGATIVE_SIGN, 0)
```
Comment (akphi) [inline packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx:438]: this expression doesn't make any sense to me

File: packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx [hunk 4]
```diff
@@ -361,13 +361,82 @@ const NumberPrimitiveInstanceValueEditor = observer(
       resetValue,
       setValueSpecification,
     } = props;
+
```
Comment (akphi) [inline packages/legend-query-builder/src/components/shared/BasicValueSpecificationEditor.tsx:364]: @MauricioUyaguari I'm a little upset how complicated this gets. I think this is not the only place and really the pain is due to us sing mobx to control the graph and the state of the component, maybe we need to think of another way to do this

**Review by akphi — CHANGES_REQUESTED** [review]

File: packages/legend-query-builder/src/stores/fetch-structure/tds/post-filter/QueryBuilderPostFilterStateBuilder.ts [hunk 5]
```diff
@@ -178,6 +179,24 @@ const processPostFilterTree = (
   postFilterState: QueryBuilderPostFilterState,
   parentPostFilterNodeId: string | undefined,
 ): void => {
+  // NOTE: This checks if the expression is a simple function expression of Minus
```
Comment (akphi) [inline packages/legend-query-builder/src/stores/fetch-structure/tds/post-filter/QueryBuilderPostFilterStateBuilder.ts:182]: Felt like this can be handled at another layer right when we process simple function expression, since here what you're doing is just to take care of them at filter/post-filter level. I'll see what I can do here

**Review by akphi — APPROVED** [review]
